src/spec-configuration/containerFeaturesOCI.ts

```diff
@@ -0,0 +1,197 @@
+export async function fetchOCIFeature(output: Log, env: NodeJS.ProcessEnv, featureSet: FeatureSet, ociCacheDir: string, featCachePath: string): Promise<boolean> {
+
+    if (featureSet.sourceInformation.type !== 'oci') {
+        output.write(`FeatureSet is not an OCI featureSet.`);
+        throw new Error('FeatureSet is not an OCI featureSet.');
+    }
+
+    const blobUrl = `https://${featureSet.sourceInformation.featureRef.registry}/v2/${featureSet.sourceInformation.featureRef.namespace}/${featureSet.sourceInformation.featureRef.featureName}/blobs/${featureSet.sourceInformation.manifest?.layers[0].digest}`;
+    output.write(`blob url: ${blobUrl}`, LogLevel.Trace);
+
+    const success = await getFeatureBlob(output, env, blobUrl, ociCacheDir, featCachePath);
+
+    if (!success) {
+        output.write(`Failed to download package for ${featureSet.sourceInformation.featureRef.featureName}`, LogLevel.Error);
+        throw new Error(`Failed to download package for ${featureSet.sourceInformation.featureRef.featureName}`);
+    }
+
+    return true;
+}
+
+export async function getFeatureManifest(output: Log, env: NodeJS.ProcessEnv, featureRef: string): Promise<OCIManifest | undefined> {
+    try {
+        const headers = {
+            'user-agent': 'devcontainer',
+            'Authorization': await getAuthenticationToken(env),
+            'Accept': 'application/vnd.oci.image.manifest.v1+json',
+        };
+
+        const options = {
+            type: 'GET',
+            url: featureRef,
+            headers: headers
+        };
+
+        const response = await request(options, output);
+        const manifest: OCIManifest = JSON.parse(response.toString());
+
+        return manifest;
+    } catch (e) {
+        output.write(`error: ${e}`, LogLevel.Error);
+        return undefined;
+    }
+}
+
+// Downloads a blob from a registry.
+export async function getFeatureBlob(output: Log, env: NodeJS.ProcessEnv, url: string, ociCacheDir: string, featCachePath: string): Promise<boolean> {
+    // TODO: Paralelize if multiple layers (not likely).
+    // TODO: Seeking might be needed if the size is too large.
+    try {
+        const tempTarballPath = path.join(ociCacheDir, 'blob.tar');
+
+        const headers = {
+            'user-agent': 'devcontainer',
+            'Authorization': await getAuthenticationToken(env),
+            'Accept': 'application/vnd.oci.image.manifest.v1+json',
+        };
+
+        const options = {
+            type: 'GET',
+            url: url,
+            headers: headers
+        };
+
+        const blob = await request(options, output);
+
+        await mkdirpLocal(featCachePath);
+        await writeLocalFile(tempTarballPath, blob);
+        await tar.x(
+            {
+                file: tempTarballPath,
+                cwd: featCachePath,
+            }
+        );
+
+        return true;
+    } catch (e) {
+        output.write(`error: ${e}`, LogLevel.Trace);
```

Comment:
I think we should always bubble up fatal errors with `LogLevel.Error`. Easier for us in the future with bug reports.